server/src/main/java/org/opensearch/common/settings/AbstractScopedSettings.java

```diff
@@ -245,10 +245,13 @@ public synchronized Settings applySettings(Settings newSettings) {
      *                  This is useful to add additional validation to settings at runtime compared to at startup time.
      */
     public synchronized <T> void addSettingsUpdateConsumer(Setting<T> setting, Consumer<T> consumer, Consumer<T> validator) {
-        if (setting != get(setting.getKey())) {
+        if (!setting.equals(get(setting.getKey()))) {
```

Comment:
I think this change alters the current behavior. 
Can we instead just update the equals check and throw exception if needed? 
```suggestion
        if (setting.getKey() != null && !setting.equals(get(setting.getKey()))) {
```

Reply:
See the original issue https://github.com/opensearch-project/opensearch-sdk-java/issues/366 which indicates that this reference equality is probably intended.

I added a comment there suggesting how to fix it (and showing an additional bug), see https://github.com/opensearch-project/opensearch-sdk-java/issues/366#issuecomment-1527791103

>  - The hard/tricky part of this fix is finding which map to use to extract the actual instance of the setting object that got registered.
>    - The registration starts at `SettingsModule.registerDynamicSetting`, but from there, there are different paths for index or node scoped settings, and each of those has two possible maps to register in.  
>      - This actually uncovers a second bug!!!  Our settings update request is only checking for settings in the ClusterSettings (node scoped) but not in the IndexScopedSettings.
>      - The equality comparison uses `get(setting.getKey())` (on the abstract parent class of both the ClusterSettings and IndexScopedSettings) which also points to a possible way of getting the object.  If you have the ClusterSettings instance you should be able to just pass the key to that `get()` method to get the setting.  (If you get null, return the "not registered" error, otherwise you're doing an update.)
>        - DO the same for the IndexScopedSettings

So start at `SettingsModule.registerDynamicSetting()` and look at the maps; you'll be extracting the `setting` from one (or both) of those maps.

Reply:
This is probably how you want to fix this in the `AddSettingsUpdateConsumerRequestHandler` class:
```java
                // Extract setting and type from writeable setting
                Setting<?> setting = extensionComponentSetting.getSetting();
                // we need to get the actual setting from nodeSetting or indexsetting maps in SettingsModule
                // use conditional based on setting properties
                Setting<?> theSettingYouNeedToPassToAddSettingsUpdateConsumer = null;
                if (setting.hasIndexScope()) {
                theSettingYouNeedToPassToAddSettingsUpdateConsumer = settingsModule.getIndexScopedSettings().get(setting.getKey());
                } else if (setting.hasNodeScope()) {                    
                    theSettingYouNeedToPassToAddSettingsUpdateConsumer = settingsModule.getClusterSettings().get(setting.getKey());
                }
                // do a null check and throw IllegalArgument exception here if neither index or node scope
                WriteableSetting.SettingType settingType = extensionComponentSetting.getType();

                // Register setting update consumer with callback method to extension
                clusterService.getClusterSettings().addSettingsUpdateConsumer(theSettingYouNeedToPassToAddSettingsUpdateConsumer, (data) -> {
```

Reply:
This is probably how you want to fix this in the `AddSettingsUpdateConsumerRequestHandler` class:
```java
// Add SettingsModule to constructor and initialize with it from ExtensionsManager

// Extract setting and type from writeable setting
Setting<?> setting = extensionComponentSetting.getSetting();
// we need to get the actual setting from nodeSetting or indexsetting maps in SettingsModule
// use conditional based on setting properties
Setting<?> theSettingYouNeedToPassToAddSettingsUpdateConsumer = null;
if (setting.hasIndexScope()) {
    theSettingYouNeedToPassToAddSettingsUpdateConsumer = settingsModule.getIndexScopedSettings().get(setting.getKey());
} else if (setting.hasNodeScope()) {                    
    theSettingYouNeedToPassToAddSettingsUpdateConsumer = settingsModule.getClusterSettings().get(setting.getKey());
}
// do a null check and throw IllegalArgument exception here if neither index or node scope
WriteableSetting.SettingType settingType = extensionComponentSetting.getType();

// Register setting update consumer with callback method to extension
if (setting.hasIndexScope()) {
    clusterService.getIndexScopedSettings()
        .addSettingsUpdateConsumer(theSettingYouNeedToPassToAddSettingsUpdateConsumer, (data) -> { ... }
}
if (setting.hasNodeScope()) {
    clusterService.getClusterSettings()
        .addSettingsUpdateConsumer(theSettingYouNeedToPassToAddSettingsUpdateConsumer, (data) -> { ... }
}                
```

Reply:
what to do if ClusterService has ClusterSettings and does not have IndexScopeSettings

Reply:
> what to do if ClusterService has ClusterSettings and does not have IndexScopeSettings

I'm not quite clear what the question is here ("has ClusterSettings").

The setting will have a list of properties, possibly including index scope and/or node scope.  (Probably not both usually, but that's not forbidden either.)

Settings with the index scope are added to IndexScopedSettings.  Settings with node scope are added to ClusterSettings.   It is possible, in theory, to be added to both, so the update consumer should be added to both.